ding/policy/pg.py

```diff
@@ -0,0 +1,210 @@
+        self._gamma = self._cfg.collect.discount_factor
+
+    def _forward_collect(self, data: dict) -> dict:
+        data_id = list(data.keys())
+        data = default_collate(list(data.values()))
+        if self._cuda:
+            data = to_device(data, self._device)
+        self._model.eval()
+        with torch.no_grad():
+            output = self._model.forward(data)
+            output['action'] = output['dist'].sample()
+        if self._cuda:
+            output = to_device(output, 'cpu')
+        output = default_decollate(output)
+        return {i: d for i, d in zip(data_id, output)}
+
+    def _process_transition(self, obs: Any, model_output: dict, timestep: namedtuple) -> dict:
+        r"""
+        Overview:
+            Generate dict type transition data from inputs.
+        Arguments:
+            - obs (:obj:`Any`): Env observation
+            - model_output (:obj:`dict`): Output of collect model, including at least ['action']
+            - timestep (:obj:`namedtuple`): Output after env step, including at least ['obs', 'reward', 'done'] \
+                (here 'obs' indicates obs after env step).
+        Returns:
+            - transition (:obj:`dict`): Dict type transition data.
+        """
+        return {
+            'obs': obs,
+            'action': model_output['action'],
+            'reward': timestep.reward,
+            'done': timestep.done,
+        }
+
+    def _get_train_sample(self, data: list) -> Union[None, List[Any]]:
+        r"""
+        Overview:
+            Get the trajectory and the n step return data, then sample from the n_step return data
+        Arguments:
+            - data (:obj:`list`): The trajectory's buffer list
+        Returns:
+            - samples (:obj:`dict`): The training samples generated
+        """
+        assert data[-1]['done'] is True, "PG needs a complete epsiode"
+
+        if self._cfg.learn.ignore_done:
+            raise NotImplementedError
+
+        R = 0.
+        for i in reversed(range(len(data))):
+            R = self._gamma * R + data[i]['reward']
+            data[i]['return'] = R
+        return get_train_sample(data, self._unroll_len)
+
+    def _init_eval(self) -> None:
+        pass
+
+    def _forward_eval(self, data: dict) -> dict:
+        data_id = list(data.keys())
+        data = default_collate(list(data.values()))
+        if self._cuda:
+            data = to_device(data, self._device)
+        self._model.eval()
+        with torch.no_grad():
+            output = self._model.forward(data)
+            if self._cfg.deterministic_eval:
+                if self._cfg.action_space == 'discrete':
+                    output['action'] = output['logit'].argmax(dim=-1)
+                elif self._cfg.action_space == 'continuous':
+                    output['action'] = output['logit']['mu']
+                else:
+                    raise KeyError("invalid action_space: {}".format(self._cfg.action_space))
```

Comment:
Why key error here, but runtime error in ding/model/template/pg.py?

Reply:
I modified runtime error in model to key error. Generally speaking, we should annotate error type in as much details as possible. runtime error is just a default case.